calyx/src/passes/remove_par.rs

```diff
@@ -0,0 +1,242 @@
+use std::collections::HashMap;
+use crate::errors;
+use crate::lang::component::Component;
+use crate::lang::{
+    ast, ast::Control, context::Context, structure::StructureGraph,
+};
+use crate::passes::visitor::{Action, Named, VisResult, Visitor};
+
+/// Pass that collapses
+/// ```
+/// (par (enable A B)
+///      (enable C D))
+///      ..)
+/// ```
+/// into
+/// ```
+/// (par (enable A B C D)
+///      ..)
+/// ```
+///
+/// when the sub-graphs induced by (enable A B) and (enable C D) have no common
+/// edges (i.e. cannot observe each other's computation).
+///
+/// For example, suppose that this were the structure graph of your component:
+/// ```
+/// ╭─╮    ╭─╮
+/// │A│    │C│
+/// ╰┬╯    ╰┬╯
+///  │      │
+///  │      │
+///  v      v
+/// ╭─╮    ╭─╮
+/// │B│    │D│
+/// ╰─╯    ╰─╯
+/// ```
+/// In this case, the program
+/// ```
+/// (par (enable A B) (enable C D))
+/// ```
+/// is equivalent to
+/// ```
+/// (par (enable A B C D))
+/// ```
+/// because there are no edges between the sub-graph induced by `A` and `B`
+/// and the sub-graph induced by `C` and `D`.
+///
+/// If instead this were your component graph:
+/// ```
+/// ╭─╮    ╭─╮
+/// │A│───>│C│
+/// ╰┬╯    ╰┬╯
+///  │      │
+///  │      │
+///  v      v
+/// ╭─╮    ╭─╮
+/// │B│    │D│
+/// ╰─╯    ╰─╯
+/// ```
+/// then `par` should be collapsed to:
+/// ```
+/// ╭─╮   ╭──╮   ╭─╮
+/// │A│──>│id│──>│C│
+/// ╰┬╯   ╰──╯   ╰┬╯
+///  │            │
+///  │            │
+///  v            v
+/// ╭─╮          ╭─╮
+/// │B│          │D│
+/// ╰─╯          ╰─╯
+/// ```
+/// which we can represented as 
+/// ```
+/// (par (enable A B C D))
+/// ```
+/// and replace the control of `enable A C` with
+/// ```
+/// (enable A id C)
+
+#[derive(Default)]
+pub struct RemovePar {
+    edge_clear: HashMap<ast::Id, Vec<ast::Id>>,
+}
+
+impl Named for RemovePar {
+    fn name() -> &'static str {
+        "remove-par"
+    }
+
+    fn description() -> &'static str {
+        "automatically merges some parallel enables"
+    }
+}
+
+// Check if two given components have an edge between them.
+fn resolve_conflicts(
+    edge_clear:&mut HashMap<ast::Id, Vec<ast::Id>>,
+    comps1: &[ast::Id],
+    comps2: &[ast::Id],
+    structure: &mut StructureGraph,
+    ctx: &Context,
+) -> Result<bool, errors::Error> {
+        let mut add_structure =
```

Comment:
is there a reason to have this instead of `resolve_conflicts`. maybe make a separate helper function?